### PR TITLE
Refactor widgets module solves #280

### PR DESCRIPTION
**Pull request by gabrielrtakeda**
# Executed tasks

### Migration
- Reducers
- Fix to relative imports
- Routes
- Tests
- Paths

### Components
- AddChoiceForm
- ChoiceCombined
- Choices
- Menu
- Page
- Match

### Containers
- ShareContainer

### Pages
- ChoicesPage
- GoalsPage

### Actions
- addMatch
- - Migration
- - Refactor action types
- - Refactor reducers

- updateMatch
- - Migration
- - Refactor action types
- - Refactor reducers

  
- Remove createOrUpdateMatch

- deleteMatch
- - Migration
- - Refactor action types
- - Refactor reducers

  
- addActivistMatch
- - Migration
- - Refactor action types
- - Refactor reducers
  
- showMatch
- - Migration
- - Refactor action types
- - Refactor reducers

## Restructure Ancestor Widget module

- Actions
  - editWidgetAsync -> async-widget-update
  - fetchWidgets -> async-widget-fetch
  - findWidget -> async-widget-select
  - Migrate widgets reducers and combine with dataExport reducers

- Containers
  - SettingsContainer

- Components
  - Widget
  - BlockTag
  - FormFooter
  - InputTag
  - OverlayWidget
  - Tag

- Pages
  - AutoFireFormPage
  - ExportWidgetData
  - DonationWidgetSettings (useless, removed)

- Organize ExportWidgetData feature
  - Migrate actions to module folder
  - Refactoring actions to use axios with thunk (promisefy)

- Selectors
  - Refactor ancestor widget selectors
  - Check best way to use selectors inside reducers

- Migration
  - Routes

## Migrate Pressure Widget plugin module directories structure

- Migration
  - Routes
  - Refact plugins routes to make it importable only once

- Actions
  - fillWidget
    - Migration
    - Refact action types
    - Refact reducers

- Components
  - PressureCount
  - PressureForm
  - TargetList
  - settings/Base
  - settings/Menu
  - Widget (pressure)

- Pages
  - EmailPage
  - FormPage

## Migrate Draft Widget plugin module directories structure

- Components
  - DraftWidget
  - DraftWidgetButton

## Migrate Form Widget plugin module directories structure

- Components
  - Form
  - Button
  - Input
  - InputForm
  - Menu

- Pages
  - Fields
  - Form

- Actions
  - addFormEntry
    - Migration
    - Refact action types
    - Refact reducers

- Migration
  - Routes

## Migrate Content Widget plugin module directories structure

- Components
  - NewEditorContentWidget
  - OldEditorContentWidget
  - ContentWidget/index

## Migrate Donation Widget plugin module directories structure

- Components
  - Donation
  - settings/Menu

- Pages
  - settings/DonationPage

- Migration
  - Routes

- Actions
  - finishTransaction
    - Migration
    - Refact action types
    - Refact reducers

### CODE REVIEW

**Comment by gabrielrtakeda** [comment]
# Widgets Module Folder Structure
The final **widgets** module folder structure was as follow:

**Folder tree of widgets module**
```
app/modules/widgets
├── __plugins__
│   ├── content
│   │   ├── action-creators
│   │   └── components
│   ├── donation
│   │   ├── action-creators
│   │   ├── components
│   │   ├── images
│   │   └── pages
│   ├── draft
│   │   └── components
│   ├── form
│   │   ├── action-creators
│   │   ├── components
│   │   └── pages
│   ├── match
│   │   ├── action-creators
│   │   ├── components
│   │   ├── containers
│   │   └── pages
│   └── pressure
│       ├── action-creators
│       ├── components
│       └── pages
├── action-creators
├── components
├── containers
├── pages
├── reducers
└── utils
```

The folder structure standard to be followed will be put into the wiki for future researches.


# Redux State Shape Standard
Another thing that is almost 100% defined is how to structure the nested redux **state shape**.
The **widgets module** redux **state shape** is nested like example below:

**Widgets module nested state shape**
```js
{
  widgets: {
    dataExport,
    list,
    plugins: {
      match,
      pressure,
      form,
      donation,
      content,
    }
  }
}
```

Each of these keys are one single **slice-reducer**, "splitted" atomically by modules e.g.:

| key            | path                                                   | type                                    |
|----------------|--------------------------------------------------------|-----------------------------------------|
| **widgets**    | app/modules/widgets/reducers/index.js                  | **root-reducer** with `combineReducers` |
| **dataExport** | app/modules/widgets/reducers/data-export.js            | **slice-reducer**                       |
| **list**       | app/modules/widgets/reducers/list.js                   | **slice-reducer**                       |
| **plugins**    | app/modules/widgets/\_\_plugins__/reducers.js          | **root-reducer** with `combineReducers` |
| **match**      | app/modules/widgets/\_\_plugins__/match/reducers.js    | **slice-reducer**                       |
| **pressure**   | app/modules/widgets/\_\_plugins__/pressure/reducers.js | **slice-reducer**                       |
| **form**       | app/modules/widgets/\_\_plugins__/form/reducers.js     | **slice-reducer**                       |
| **donation**   | app/modules/widgets/\_\_plugins__/donation/reducers.js | **slice-reducer**                       |
| **content**    | app/modules/widgets/\_\_plugins__/content/reducers.js  | **slice-reducer**                       |

File: app/modules/mobilizations/blocks/components/specs/block.spec.js [hunk 1]
```diff
@@ -367,7 +367,7 @@ describe('Block', () => {
     })
   })
 
-  describe.skip('#displayDropDownMenu', () => {
```
Comment (lpirola) [inline app/modules/mobilizations/blocks/components/specs/block.spec.js:370]: Testes de volta! o/

**Review by lpirola — COMMENTED** [review]
Podemos fazer o merge sem problemas. Testamos melhor no branch refactor-modules.

File: app/modules/mobilizations/blocks/components/specs/block.spec.js (hunk 1 above)
Comment (gabrielrtakeda) [reply app/modules/mobilizations/blocks/components/specs/block.spec.js:370]: @lpirola Infelizmente não. O `.skip` foi apenas migrado pro describe pai..
```patch
 -describe('Block', () => {
 -  describe.skip('#constructor', () => {
 +describe.skip('Block', () => {
 +  describe('#constructor', () => {
```

**Review by gabrielrtakeda — COMMENTED** [review]
Explain block widget tests ignore by the parent describe.

**Pull request merged**